### PR TITLE
feat: adds workflow for pretraining a super-network

### PR DESCRIPTION
**Pull request by aaronkl**
#### Reference Issues/PRs
#157 

#### What does this implement/fix? Explain your changes.
add pretrain_super_network.py + unittest to enable super-network training from scratch following litgpt workflow

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the
terms of your choice.

### CODE REVIEW

File: whittle/pretrain_super_network.py [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+    out_dir: Path,
+    tokenizer_dir: Optional[Path],
+    train: TrainArgs,
+    eval: EvalArgs,
+    training_strategy: BaseTrainingStrategy,
+) -> None:
+    model = state["model"]
+    optimizer = state["optimizer"]
+
+    if eval.initial_validation:
```
Comment (rheasukthanker) [inline whittle/pretrain_super_network.py:229]: Do we want to simply use litgpt's validation loop? Do we want to reset super network before every validation? Ideally would be nice to report : smallest, medium and largest architecture loss for a more robust validation

File: whittle/pretrain_super_network.py [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+        val_loss = validate(fabric, model, val_dataloader, max_iters=eval.max_iters)
+        val_loss = f"{val_loss:.3f}"
+    else:
+        fabric.print("Verifying settings ...")
+        validate(
```
Comment (rheasukthanker) [inline whittle/pretrain_super_network.py:234]: Same as above comment for validation

File: whittle/pretrain_super_network.py [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+            metrics.update(throughput_metrics)
+            fabric.log_dict(metrics, step=state["iter_num"] - 1)
+
+        if (
+            val_dataloader is not None
+            and not is_accumulating
+            and state["step_count"] % eval.interval == 0
+        ):
+            t0 = time.perf_counter()
+            val_loss = validate(fabric, model, val_dataloader, max_iters=eval.max_iters)
```
Comment (rheasukthanker) [inline whittle/pretrain_super_network.py:370]: same as above comment for validation

File: whittle/pretrain_super_network.py [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+            save_checkpoint(
+                fabric,
+                state,
+                tokenizer_dir,
+                out_dir / f"step-{state['step_count']:08d}" / "lit_model.pth",
+            )
+
+    # Final validation
+    if eval.final_validation:
+        val_loss = validate(fabric, model, val_dataloader, max_iters=eval.max_iters)
```
Comment (rheasukthanker) [inline whittle/pretrain_super_network.py:395]: same as above comment for validation

File: whittle/pretrain_super_network.py [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+
+        is_accumulating = (
+            state["iter_num"] % train.gradient_accumulation_iters(devices) != 0
+        )
+        with fabric.no_backward_sync(model, enabled=is_accumulating):
+            loss = training_strategy(
+                model,
+                input_ids,
+                targets,
+                scale_loss=1 / train.gradient_accumulation_iters(devices),
```
[the remark this answers is not in the document]
Comment (aaronkl) [reply whittle/pretrain_super_network.py:301]: good catch!

File: whittle/pretrain_super_network.py (hunk 1 above)
Comment (aaronkl) [reply whittle/pretrain_super_network.py:229]: You are right we should definitely reset the super-network. I am not sure about evaluating the smallest, medium and largest arch, though. I agree that it is convenient for diagnosis, but it also adds an additional overhead. Maybe we could have different callbacks to switch between different validation themes? Could be a separate PR.

**Review by rheasukthanker — APPROVED** [review]
